### ISSUE TITLE
The fully qualified file name must be less than 260 characters, and the directory name must be less than 248

### ISSUE THREAD

**Issue by PrimeObjects** [Microsoft/dotnet#234]
I have moved my project folder to C:\V7 a short one, I keep getting the error below when trying to publish my web app to Azure web app. Can the message below at least show whitch file has long file name?

How to solve this issue?

Thanks
Gary

Microsoft .NET Development Utility CLR-x86-1.0.0-beta7-15019
Copying to output path C:\Users\gzhan_000\AppData\Local\Temp\PublishTemp\PrimeObjects.Platform.Web65

System.IO.PathTooLongException: The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.
   at System.IO.Path.NormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
   at System.IO.Path.GetDirectoryName(String path)
   at Microsoft.Dnx.Tooling.Publish.PublishOperations.ExtractFiles(ZipArchive archive, String targetPath, Func`2 shouldInclude)
   at Microsoft.Dnx.Tooling.Publish.PublishOperations.ExtractNupkg(ZipArchive archive, String targetPath)
   at Microsoft.Dnx.Tooling.NuGetPackageUtils.ExtractPackage(String targetPath, FileStream stream)
   at Microsoft.Dnx.Tooling.NuGetPackageUtils.<>c__DisplayClass0_0.<<InstallFromStream>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dnx.Runtime.Internal.ConcurrencyUtilities.<ExecuteWithFileLocked>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dnx.Tooling.NuGetPackageUtils.<InstallFromStream>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dnx.Tooling.Packages.AddCommand.<Execute>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dnx.Tooling.Publish.PublishPackage.Emit(PublishRoot root)
   at Microsoft.Dnx.Tooling.Publish.PublishRoot.Emit()
   at Microsoft.Dnx.Tooling.Publish.PublishManager.Publish()
   at Microsoft.Dnx.Tooling.PublishConsoleCommand.<>c__DisplayClass0_0.<Register>b__1()
   at Microsoft.Dnx.Runtime.Common.CommandLine.CommandLineApplication.Execute(String[] args)
   at Microsoft.Dnx.Tooling.Program.Main(String[] args)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Dnx.Runtime.Common.EntryPointExecutor.Execute(Assembly assembly, String[] args, IServiceProvider serviceProvider)
   at Microsoft.Dnx.Host.Bootstrapper.RunAsync(List`1 args, IRuntimeEnvironment env, FrameworkName targetFramework)
   at Microsoft.Dnx.Host.RuntimeBootstrapper.ExecuteAsync(String[] args, FrameworkName targetFramework)
   at Microsoft.Dnx.Host.RuntimeBootstrapper.Execute(String[] args, FrameworkName targetFramework)
   at Microsoft.DNX.Tasks.Dnu.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

**Comment by oxishmit** [comment]
Long Path Tool can solve that problem and it is free, try it

**Comment by treestryder** [comment]
That error will also fire when there are more than X separators, which includes periods, as well as slashes. Sorry, can't recall what  X is and can't test it just now.

**Comment by qbikez** [comment]
For me the workaround was to set `PublishTempFolderName` property in `.pubxml` publishing profile to some short name:

```
<PublishTempFolderName>pubtmp</PublishTempFolderName>
```

This way DNU will use `%UserProfile%\pubtmp` path, which was enough to fix the problem for me. If you want to investigate further, look at `C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\Web\Microsoft.DNX.Publishing.targets` file, where these properties are defined.

**Comment by NightkingCH** [comment]
Is there any reason why we can't define a full path for the temp folder? That's just annoying... When I try to specify "C:\PublishTemp" I get following error "The given path's format is not supported.". Sure when I concat the string like that, it doesn't suprise me. 

I mean we have 2016 and we have still to deal with the "260 characters"-error. So let me decide where I can put my stuff...

Is there any way to NOT use the "%UserProfile%"-Path? Do we really need to hard code ```[System.IO.Path]::GetTempPath()``` within the msbuild task? Why not create a default variable which we can override?

My workaround for now:

1. Create a new user account for the build agent
2. Change his TEMP & TMP environment variable to a short path (C:\PubTmp)
3. Register the build agent as a windows service and use the created user account
4. Change the nodejs global prefix to a short path (C:\nodejs) (only if you are using gulp)

Afterwards the build runs through with the specified publishing profile.

**Comment by domyorker** [comment]
I've had nothing but headaches with .NET Core...including the above issue with file names reportedly too long, VS crashing while loading NPM modules, and the build failing because my references are in the 4.x DNX and not in DNX 5. No wonder so many companies stick with "legacy" systems/software. Perhaps major software developers are so caught up in the hype of JS, Open Source and universality that we end up with unfinished (rushed to the public) systems. It is so disappointing.

**Comment by richlander** [comment]
Please re-activate if this is still an issue.

**Comment by Seankps** [comment]
Issue still exists in 2017 in VS2017

**Comment by richlander** [comment]
@Seankps Can you share some diagnostic information or a repro with me? If it is private, feel free to send it to rlander@ms.

**Comment by Seankps** [comment]
@richlander not really. It's the same error. A file path of 260 characters should be acceptable in this age, on Windows 10, in Visual Studio 2017. This is a ridiculous limitation to have still. Like some serious legacy stuff that someone avoided addressing for too long. 

Or there should be a work around better than shortening the path if you can. I understand that this is probably more of a Windows issue than a .NET issue. As a developer I just don't understand how they got away without fixing something like this for so long

**Comment by richlander** [comment]
Agreed.

Is it NPM packages that hit this in your scenario or something else? I'm asking because it's easier to describe why this should be fixed if I can provide some more detail on concrete cases.

**Comment by Seankps** [comment]
@richlander I think so. But can't be sure because the VS build error doesn't specify the file/location

**Comment by NightkingCH** [comment]
As @Seankps said, it still exists and we have still the workaround in our build environments. It's not only related to npm packages. Regardless what triggers the error it should be possible to define where we want to store our stuff. I agree with that in the most cases it would work. But that does not mean we have to ignore the other 10%-cases. I mean, it's not your fault, if we have paths longer than 260 characters, this is a restriction we have to deal with, but give us an option to short our paths.

So whats about my idea? Defining a variable with the default value, I don't mind if it's the users temp path, and then let us override the variable with a build argument?

**Comment by csnellemann** [comment]
> Can the message below at least show which file has long file name?

@richlander We second the above request from @PrimeObjects. The specific path needs to be available somehow. We experience the issue when our build server tries to get our internal NuGet packages with NuGet install. The error message only tells us which package caused the problem, not which path (it prints a source URL, but that is much shorter than the limit). Since any number of temp file/folder names may be involved during the NuGet install, finding ways to shorten the path turns into very painful trial and error.

**Comment by derekpec** [comment]
While I agree that the filepath should not be limited to 260 characters, it's important that any underlying bug that may be causing unnecessarily long filepaths is not swept under the rug.  I'm experiencing a similar issue and have narrowed it down to the same as described by @pellelauritsen in https://github.com/aspnet/Home/issues/790: 

> First publish is fine, but second time I get a bin-folder containing my views in a couple of directories down Release\Views. If I publish yet again I get a new set inside so now the bin-folder structure is \bin\Release\bin\Release\Views. This goes on until VS cant build the project because of the above mentioned error!

**Comment by danmosemsft** [comment]
I added the actual path to the exception string (in most cases) https://github.com/dotnet/coreclr/commit/6ba74dc2a7194f8d6c86c3aeab572a074ef645c8 but this change is not in a released build.
Can you try using Process Explorer to see if the long path shows up there?

**Comment by JeremyKuhne** [comment]
@derekpec, you resolved the underlying [issue](https://github.com/Microsoft/msbuild/issues/2185). @danmosemsft added the paths to the messages where we didn't have them in Core. Is there anything more we can do with this issue?

**Comment by richlander** [comment]
.NET Core users can temporarily install and target .NET Core 2.1 preview builds. This build has the diagnostic changes in it that @danmosemsft mentioned. When you use this build, it will report back which paths have the long file path issue. This info will be tremendously helpful to us.

**Comment by ashley200** [comment]
Long path tool can definitely solve this problem. I am sure.

**Comment by ashley200** [comment]
Try Long Path tool.

**Comment by chiefengineer** [comment]
To add grist to the mill...

I'm getting the following error when attempting to install a nuget package into a standard .NET framework 4.7 project:

The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.

I'm using Visual Studio 2017 15.3.3 Enterprise (latest and greatest), this is a .NET Framework 4.7 (lest someone confuse this with .NET "Standard" or .NET "Core", this is just the regular 'ol framework that's been around for the last 15 years or so), latest and greatest nuget.exe and a homegrown nuget server that's been working just fine since I recompiled it with .NET 4.7 when 4.7 hit the streets (January or so, I think) that is just a home for nuget.server, version 2.11.3.0 build iaw the nuget.server's author's instructions, all running on the latest and greatest Windows - 10 creator (1703).

Given that this is my package, I have full control over the source code and the source to all of the ancestors. No other third party packages in there, just my stuff and MS's. The interesting part is that I have used this package in the past with no changes in the name, but for this go round, I rebuilt it to add a feature (a new method in an already existing class) and am now getting this error.

Even more interesting is that I have packages from the same library, with the same namespace conventions, with **longer** names, that work just fine and have installed into this same project with no problems at all.

I have already tried shrinking the package name, shrinking the class names within the package itself, cleaning out the build directory, cleaning out the package home from the nuget server (it's a local server with the latest nuget.server installed which otherwise works just fine), and even clearing out the bin directory of the project in question, clearing out ALL the bin directories of ALL the ancestors to the "offending" package, clearing out the package cache, rebooting the computer and rebuilding the entire nuget package chain from scratch, all to no avail.

The short version is that I started with as close to "File | New" as one can get and am getting this failure with this one peculiar package.

A little light shed on where to look for offending things would be of great assistance.

@ashley200 - wonderful - would you mind explaining, in detail for those of us that have never heard of this tool, just *how* one would go about applying this tool to "solve" the problem?

**Comment by danmosemsft** [comment]
@chiefengineer have you tried sysinternals process monitor to record file system activity? It may show the path that is causing the error.

**Comment by chiefengineer** [comment]
Not to put too fine a point on it, but "huh?". I didn't think of that. I'll
go investigate and see what comes back. If you have the time to shed a
little more light on what you are referencing, that'd be awesome as I am
working in the dark here.
Either way, thanks for the suggestion!


"Bang the Button - It Just Works!"

**Comment by chiefengineer** [comment]
OK, big Thank You  @danmosemsft who suggested digging around with the SysInternals process monitor. After fiddling with it for a bit, I finally figured out how to narrow the result set to just file activity. What I noticed, and the nuget engineers should TAKE NOTE of this: The problem was NOT a too-long project name, rather, nuget was attempting to update a package that was no longer there. Why it went away is a mystery yet to be solved. I normally stay out of the packages directory and do not fuss with the packages.config file. I *think* that this might have to do with my impatience waiting for VS to start, load all the goodies and then allow me to perform a "Manage NuGet packages" - update all. Rather than wait until VS has settled down, I pushed it a bit (hey, the buttons responded so there *shouldn't* be any problems...). 

As a result, the packages directory was absolutely chock full of old stuff that did NOT belong there. So, I manually cleaned up all the chaff, manually cleaned up the packages.config file, restarted VS, waited for it to settle down, performed my NuGet updates and viola! no problem - HAVING NOT CHANGED ANY OF THE ANCESTRAL PACKAGE NAMES BY EVEN A SINGLE CHARACTER.

So, what do I conclude from this? It is my belief, and the guys that actually build nuget and nuget.server, should take a closer look at the errors being thrown, and I *think* that the error is not so much a path too long error, rather it's a "hey, I didn't find the file I expected so the file name is full of junk (and probably too long now) so I'll throw an error that says it's too long and quit. It's a failure to handle a missing package/package directory that is causing this particular problem

I solved my problem by ensuring that all the package directories were clean of all junk and rebuilding from a clean source. My problem is now solved.

**Comment by danmosemsft** [comment]
@chiefengineer glad to hear it! Would you open an issue in the nuget github repo to help them fix whatever it was?

**Comment by chiefengineer** [comment]
Sure - there are a couple of bugs that I have noticed that need addressing,
I'll do that later today.


"Bang the Button - It Just Works!"

**Comment by chiefengineer** [comment]
...and done. Thanks again for your help.


"Bang the Button - It Just Works!"

**Comment by danmosemsft** [comment]
You're welcome.

**Comment by kriscoleman** [comment]
@ashley200 I don't know of a way to use Long Path Tool in an automated build process. I encountered this error during a TFS CI build, because NPM was pulling down many packages and the file paths got too long. I don't think I could automate Long Path Tool into my build process to fix this.

**Comment by Petermarcu** [comment]
Looks like everything is understood here and issues are filed elsewhere. Can we close this one here?

**Comment by chiefengineer** [comment]
As long as the error is understood, documented elsewhere and in the queue
to be fixed, sure. This one caused me all sorts of headaches!


"Bang the Button - It Just Works!"

**Comment by Petermarcu** [comment]
@chiefengineer opened the long file path issue and it got closed as a duplicate of this one. https://github.com/NuGet/Home/issues/3324. I think thats everything from this issue so I'm going to close.

**Comment by surhere** [comment]
I solved my problem by ensuring that all the package directories were clean of all junk and rebuilding from a clean source. My problem is now solved. It was a custom js folder. 💯 

Copying file Content\js\lib\datatables\cdnjs.cloudflare.com\ajax\libs\jqueryui-touch-punch\0.2.3\jquery.ui.touch-punch.min.js to obj\Debug\Package\PackageTmp\Content\js\lib\datatables\cdnjs.cloudflare.com\ajax\libs\jqueryui-touch-punch\0.2.3\jquery.ui.touch-punch.min.js failed. The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.

**Comment by JillGoodman** [comment]
@chiefengineer - thanks for the great explanation.  i've been knocking my head for a few days!

**Comment by Legends** [comment]
Usually I keep my VS projects in the default folder which is under documents.
As my C: drive is running low on space, I decided to move the documents folder to: `"D:\system\Documents"`.

Default folder path before: `C:\Users\User\Documents`.

The current path `"D:\system\Documents"`. is shorter than the default path, but nevertheless I get this exception.

**Comment by Legends** [comment]
Actually this is the problem not my folder path:

![image](https://user-images.githubusercontent.com/966368/70150484-5c7a6d00-16aa-11ea-9cd5-aebab9e96601.png)

**Comment by Bosch-Eli-Black** [comment]
We ran into this error today. (Please note that I've replaced the names of the projects with "abc.....", to hide the real identity of the project. However, the masked names are _exactly_ the same lengths as the original names)


Severity | Code | Description | Project | File | Line | Suppression State
-- | -- | -- | -- | -- | -- | --
Error | MSB3491 | Could not write lines to file   "obj\Debug\netstandard2.0\abcde.fgh.ijklmn.opq.rstuvwxyzabcdefgh.GeneratedMSBuildEditorConfig.editorconfig".   Path:   obj\Debug\netstandard2.0\abcde.fgh.ijklmn.opq.rstuvwxyzabcdefgh.GeneratedMSBuildEditorConfig.editorconfig   exceeds the OS max path limit. The fully qualified file name must be less   than 260 characters. | abcde.fgh.ijklmn.opq.rstuvwxyzabcdefgh | C:\Program Files (x86)\Microsoft Visual   Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.Managed.Core.targets | 153 |  

In this case I'm thinking that this is just a Visual Studio caching issue and that deleting the `obj\Debug` folder should fix the issue, but I thought I'd still report it.

**Comment by mldisibio** [comment]
Deleted both \bin and \obj folders and still receive this error:
`Error MSB3491 Could not write lines to file "obj\Debug\netcoreapp3.1\staticwebassets\<my.webapi.project>.StaticWebAssets.Manifest.cache". Path: obj\Debug\netcoreapp3.1\staticwebassets\<my.webapi.project>.StaticWebAssets.Manifest.cache exceeds the OS max path limit. The fully qualified file name must be less than 260 characters.	incite.logretriever.logfiles.webapi	C:\Program Files\dotnet\sdk\5.0.103\Sdks\Microsoft.NET.Sdk.Razor\build\netstandard2.0\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets	175`
Also, this is a simple 3.1 web api project, no Razor or static content outside of Swagger, although I realize Razor can be part of the templates.

**Comment by danmoseley** [comment]
Hello, for new variations of this issue please open a fresh issue in a more targeted repo. The dotnet/SDK or aspnetcore repos may be most relevant. Or the Visual Studio feedback tool if it is VS specific. Or Nuget repo.

**Comment by danmoseley** [comment]
@mldisibio I would use "process monitor" tool to see what the actual path is causing problems.

**Comment by mldisibio** [comment]
@danmoseley  Thanks Dan. Yes I should open a new issue, when I get some time to document all the details correctly. Right now working against some deliverable deadlines. Meanwhile thought I would post, for general information, one possible fix as [documented by MS ](https://docs.microsoft.com/en-us/xamarin/android/troubleshooting/questions/path-too-long-exception#fix). Add to project file:
`<IntermediateOutputPath>C:\Projects\MyShortPath</IntermediateOutputPath>` solved the immediate problem.
The other proposed element: `<UseShortFileNames>True</UseShortFileNames>` did not seem to work for me.

**Comment by danmoseley** [comment]
Thanks for sharing that info @mldisibio

**Comment by hrumhurum** [comment]
Just hit this issue today:

> ... .GeneratedMSBuildEditorConfig.editorconfig exceeds the OS max path limit. The fully qualified file name must be less than 260 characters.

 Environment: Windows Version 20H2 (OS Build 19042.985), Visual Studio 16.10.0

Setting `Enable Win32 long paths` policy in Group Policy editor produced a working workaround.

1. Click the `Window` key and type `gpedit.msc`, then press the `Enter` key. This launches the Local Group Policy Editor
2. Navigate to `Local Computer Policy` > `Computer Configuration` > `Administrative Templates` > `System` > `Filesystem`
3. Double click on `Enable NTFS long paths`
4. Set the policy state to `Enabled`
5. Click `OK`

Reboot the machine and bingo, that error is gone.

**Comment by danmoseley** [comment]
Thank you for sharing that @hrumhurum

**Comment by VannakSaing** [comment]
> > Just hit this issue today:
> > > ... .GeneratedMSBuildEditorConfig.editorconfig exceeds the OS max path limit. The fully qualified file name must be less than 260 characters.
> > 
> > 
> > Environment: Windows Version 20H2 (OS Build 19042.985), Visual Studio 16.10.0
> > Setting `Enable Win32 long paths` policy in Group Policy editor produced a working workaround.
> > 
> > 1. Click the `Window` key and type `gpedit.msc`, then press the `Enter` key. This launches the Local Group Policy Editor
> > 2. Navigate to `Local Computer Policy` > `Computer Configuration` > `Administrative Templates` > `System` > `Filesystem`
> > 3. Double click on `Enable NTFS long paths`
> > 4. Set the policy state to `Enabled`
> > 5. Click `OK`
> > 
> > Reboot the machine and bingo, that error is gone.
> 
> It is not working for me (Version 20H2 Os Build 19042.1526) vs 2019

all error gone by your following step.

**Comment by bojkel** [comment]
Issue still exists in January, 2023.

Windows 10
Visual Studio 17.4.4
.NET Core SDK 6.0.201 & 7.0.102 (from Visual Studio)

**Comment by bojkel** [comment]
Initially I tried following @hrumhurum's steps, but I didn't have the option to enable NTFS long paths from there.
What fixed it for me is enabling Windows to accept long paths from the Registry Editor.
Here are the steps:

1. Open Registry Editor as Administrator (windows key, type `regedit` and open as Administrator)
2. Go to `HKEY_LOCAL_MACHINE/SYSTEM/CurrentControlSet/Control/FileSystem`
3. Double click on `LongPathsEnabled`
4. Set the value to 1
5. Click "OK"
6. Restart your machine

**Comment by anafletcher88** [comment]
Has anyone used LongPath Tool?

**Comment by ErikBrown2** [comment]
> Has anyone used LongPath Tool?

Yes
